File: src/observable/observable_list.ts
import { create_message } from "../comm/__init__.js";
import { ListChangeMessage } from "./messages.js";
import { ObservableBase, ObservableKwds } from "./observable_base.js";
import { ObservableListSchema } from "./schema.js";


export class ObservableList<T> extends ObservableBase<ObservableListSchema<T>, ListChangeMessage<T>> {
    #list: T[]

    private constructor(list: T[], kwds: ObservableKwds = {}) {
        super(kwds);
        this.#list = Array.from(list);
    }


    static create<T>(list: T[], kwds?: ObservableKwds): ObservableList<T> {
        return new ObservableList(list, kwds);
    }

    static from_schema<T>(schema: ObservableListSchema<T>): ObservableList<T> {
        return ObservableList.create(schema.list, { id: schema.data_id, remote: true });
    }

    get list(): readonly T[] {
        return Array.from(this.#list);
    }

    append(item: T): void {
        this.#emit_changes({ [this.#list.length]: item }, []);
    }

    pop(index: number = -1): T {
        const item = this.#list[index];
        this.#emit_changes({}, [index]);
        return item;
    }

    extend(items: T[]): void {
        const start = this.#list.length;
        const inserts: { [index: number]: T } = {};
        items.forEach((item, index) => {
            inserts[start + index] = item;
        });
        this.#emit_changes(inserts, []);
    }

    clear(): void {
        const removes = this.#list.map((v, i) => i);
        this.#emit_changes({}, removes);
    }

    insert(index: number, item: T): void {
        this.#emit_changes({ [index]: item }, []);
    }

    remove(index: number): void {
        this.#emit_changes({}, [index]);
    }

    map<R>(fn: (v: T) => R): ObservableList<R> {
        const list = this.#list.map(fn);
        const result = ObservableList.create(list, { remote: false });
        this.watch(change => {
            const inserts: { [index: number]: R } = {};
            for (let index in change.inserts) {
                inserts[index] = fn(change.inserts[index]);
            }
            result.#apply_changes(inserts, change.removes);
            result.#emit_changes(inserts, change.removes, true);
        });
        this.network?.register(result);
        return result;
    }

    // /**
    //  * A {@link TransformFn<any, T>} that coerces any value to {@link T}.
    //  * Defaults to the constructor function of {@link T}.
    //  *
    //  * @type {TransformFn<any, T>}
    //  */
    // coerce: TransformFn<any, T>;


    protected _handle_change_message(msg: ListChangeMessage<T>): void {
        this.#apply_changes(msg.inserts, msg.removes);
    }

    #emit_changes(
        inserts: { [index: number]: T },
        removes: number[],
        local: boolean = false,
    ): void {
        const message = create_message<ListChangeMessage<T>>(
            {
                "data_id": this.id,
                "inserts": inserts,
                "removes": removes,
            },
        )
        this._emit_change_message(message, local);
    }

    #apply_changes(
        inserts?: { [index: number]: T },
        removes?: number[],
    ): void {
        if (removes) {
            removes = removes.map(
                index => index < 0 ? this.#list.length + index : index
            );
            for (let index of removes.sort((a, b) => b - a)) {
                this.#list.splice(index, 1);
            }
        }

        if (inserts) {
            for (let index_str in inserts) {
                const item = inserts[index_str];
                let index = Number(index_str);
                if (index < 0) {
                    index += this.#list.length;
                }
                this.#list.splice(index, 0, item);
            }
        }
    }

};

File: src/observable/observable_list.py
from typing import Any, Callable, cast

from .schema import ObservableListSchema

from ..comm import create_message
from .messages import ListChangeMessage
from .observable_base import ObservableBase

__all__ = ["ObservableList"]


def or_default[T](v: T | None, d: T) -> T:
    if v is None:
        return d
    return v


class ObservableList[T](ObservableBase[ObservableListSchema[T], ListChangeMessage[T]]):
    __list: list[T]
    __coerce: Callable[[Any], T] | None
    __validate: Callable[[T], T] | None

    def __init__(
        self,
        *,
        coerce: Callable[[Any], T] | None = None,
        validate: Callable[[T], T] | None = None,
    ):
        super().__init__()
        self.__list = []
        self.__coerce = coerce
        self.__validate = validate

    # Dunder methods for list-like syntax
    def __len__(self) -> int:
        return len(self.__list)

    def __getitem__(self, index: int | slice):
        if isinstance(index, slice):
            return self.__list[index.start : index.stop : index.step]
        return self.__list[index]

    def __setitem__(self, index: int | slice, value: T | list[T]):
        if isinstance(index, slice):
            value = cast(list[T], value)
            self.__apply_slice(index, value)
        else:
            value = cast(T, value)
            self.__apply_changes(inserts={index: value}, removes=[index])

    def load_schema(self, schema: ObservableListSchema[T]) -> None:
        self[:] = schema["list"]

    def __json__(self) -> ObservableListSchema[T]:
        return {"data_id": self.id, "list": self.__list}

    # Higher-level methods
    def append(self, item: T):
        self.__apply_changes(inserts={len(self.__list): item})

    def pop(self, index: int = -1):
        item = self.__list[index]
        self.__apply_changes(removes=[index])
        return item

    def extend(self, items: list[T]):
        start = len(self.__list)
        self.__apply_changes(inserts={start + i: item for i, item in enumerate(items)})

    def clear(self):
        self.__apply_changes(removes=list(range(len(self.__list))))

    def insert(self, index: int, item: T):
        self.__apply_changes(inserts={index: item})

    def remove(self, index: int):
        self.__apply_changes(removes=[index])

    def _handle_client_change(self, change: ListChangeMessage[T]) -> None:
        # return super()._handle_client_change(change)
        self.__apply_changes(change["inserts"], change["removes"])

    def __apply_slice(self, slice_obj: slice, items: list[T]):
        start = or_default(slice_obj.start, 0)
        stop = or_default(slice_obj.stop, len(self.__list))
        step = or_default(slice_obj.step, 1)
        indices_to_remove = list(range(start, stop, step))
        indices_to_insert = list(
            range(start, start + len(items) * abs(step), abs(step))
        )

        if step < 0:
            indices_to_remove.reverse()
            indices_to_insert.reverse()

        inserts = {index: item for index, item in zip(indices_to_insert, items)}
        self.__apply_changes(inserts=inserts, removes=indices_to_remove)

    def __apply_changes(
        self,
        inserts: dict[int, T] | None = None,
        removes: list[int] | None = None,
    ):
        message = create_message(
            ListChangeMessage[T],
            {
                "data_id": self.id,
                "inserts": {},
                "removes": [],
            },
        )

        new_list = [*self.__list]
        if removes:
            for index in sorted(removes, reverse=True):
                # self.__list.pop(index)
                new_list.pop(index)
            message["removes"] = removes

        if inserts:
            for index, item in sorted(inserts.items()):
                # self.__list.insert(index, item)
                new_list.insert(int(index), item)
            message["inserts"] = inserts

        self.__list = new_list
        if self.network is not None:
            self.network.emit_change(self, message)
